### ISSUE TITLE
ufw blocks non-standard ports, requiring manual intervention when running without a domain

### ISSUE THREAD

**Issue by pierremonico** [dokku/dokku#3455]
## Description of problem
I deployed Dokku on a Digital Ocean Droplet using their one click install. 
Setup went fine, and the remote was added. 
At very first push, I end up with the message:
```
 ! [remote rejected] master -> master (pre-receive hook declined)
error: failed to push some refs to 'dokku@myIP:appname'
```
I checked the events.log file and these are the last lines:
```
recognizer dokku-installer.py[9094]: Stopping nginx (via systemctl): nginx.service.
recognizer dokku-installer.py[9094]: Starting nginx (via systemctl): nginx.service.
recognizer dokku-installer.py[9094]: /bin/sh: 1: stop: not found
recognizer dokku[15617]: -----> Restarting in parallel
recognizer dokku[15617]:  !     App appname has not been deployed
```

I destroyed the droplet and started again, but nothing helps. 
I have the app running on Heroku so the settings should sensibly be the same.

### How reproducible
See above.

### Steps to Reproduce

1. Run the one click installation of Dokku on Digital Ocean
2. Configure Dokku and create the app on the host
3. Add the remote to the local project
4. `git push dokku master`

#### Actual Results
Error while deploying.

#### Expected Results
Successful deployment.

## Environment Information
- Flask app 
- requirements.txt wit all dependencies 
- Procfile with `web: gunicorn appname:app`
- runtime.txt with `python-3.7.2`

### `dokku report APP_NAME` output
[log.txt](https://github.com/dokku/dokku/files/2912584/log.txt)

### How (deb/make/rpm) and where (AWS, VirtualBox, physical, etc.) was Dokku installed?:
Digital Ocean

**Comment by pythdasch** [comment]
same here with a Django app. which work for our production but We try to create a new app for a test server and then deployment fails

Dokku version 0.11.

**Comment by pierremonico** [comment]
As a side note, I made a third try after updating dokku to the latest version (0.14.6) but that didn't help. 
Btw, I also had to use `wget -nv -O - https://packagecloud.io/dokku/dokku/gpgkey | apt-key add -` like specified [here](https://github.com/dokku/dokku/issues/3390#issuecomment-453178841) since I had the same error as [here](https://github.com/dokku/dokku/issues/3421#issue-403014129).

**Comment by pythdasch** [comment]
@pierremonico did this solved your issue?

**Comment by pierremonico** [comment]
(@pythdasch no.)

Here also the few last lines with `trace`enabled:
```
+ for script in $PLUGIN_ENABLED_PATH/*/commands
+ set +e
+ /var/lib/dokku/plugins/enabled/tags/commands git-receive-pack ''\''recognizer'\'''
+ exit_code=10
+ set -e
+ [[ 10 -eq 10 ]]
+ continue
+ for script in $PLUGIN_ENABLED_PATH/*/commands
+ set +e
+ /var/lib/dokku/plugins/enabled/tar/commands git-receive-pack ''\''recognizer'\'''
+ exit_code=10
+ set -e
+ [[ 10 -eq 10 ]]
+ continue
+ [[ 1 -eq 0 ]]
+ set +f
+ exit 0
To IP:appname
 ! [remote rejected] master -> master (pre-receive hook declined)
error: failed to push some refs to 'dokku@IP:appname'
```

**Comment by josegonzalez** [comment]
Can you show the _entire_ output of a trace-enabled deploy?

**Comment by pierremonico** [comment]
Solved it seconds ago by increasing swapfile size:
```
sudo swapoff -a
sudo dd if=/dev/zero of=/swapfile bs=4G count=102
sudo mkswap /swapfile
sudo swapon /swapfile
```

**Comment by pierremonico** [comment]
Although deployment worked, the app is still not responding to requests and my error logs don't show anything significant. I even redeployed to a droplet with more memory, but even after successful deployment I can't access the app through HTTP. 
It works on Heroku and I am also able to perfectly access it locally running the Flask development server or Gunicorn. I suspect there is something wrong with the DO one-click install and or the Python buildpack. 
Should that be another issue?

**Comment by pierremonico** [comment]
And finally got everything working by reading [this post](https://www.digitalocean.com/community/questions/security-tips-and-suggestion-dokku-node-js-app). You need to open the appropriate port using ufw. 

After all the hassle, I am not sure if Digital Ocean can really call that a one-click install :) 

Sorry if this is not the right place for all of this, but I wish I would have found it when trying to solve the problems. Thanks for making Dokku.

**Comment by josegonzalez** [comment]
The big problem here is that we don't integrate with `ufw`, which is enabled by default on newer versions of Ubuntu. There are a few routes here:

- the web installer (and whatever replaces it) detects ufw and notifies the user of steps they may need to take
- some documentation around port management to output to users that they'll need to open the port appropriately
- log a warning for port-based deploys to have users disable ufw/firewalls (AWS security groups are another example of where this sucks)
- opt-in support for ufw (or an official, non-core plugin to handle it)

The first three are easy, the last one is harder.

**Comment by josegonzalez** [comment]
I'm going to do the first two. For the last one, I think if this comes up again, we can revisit the issue.

**Comment by josegonzalez** [comment]
Closing as there is a pull request open.
